File: src/tak/ptn.rs
use crate::tak::{TakAction, TakDir, TakGameState, TakPlayer, TakPos, TakVariant, TakWinReason};

pub fn action_to_ptn(action: &TakAction) -> String {
    match action {
        TakAction::Place { pos, variant } => {
            format!("{}{}", variant_to_string(variant), pos_to_string(pos))
        }
        TakAction::Move { pos, dir, drops } => {
            let drops_str: String = drops.iter().map(|d| d.to_string()).collect();
            let drops_str = if drops_str.len() == 1 {
                "".to_string()
            } else {
                drops_str
            };
            let take_num = drops.iter().sum::<u32>();
            let take_str = if take_num == 1 {
                "".to_string()
            } else {
                take_num.to_string()
            };
            format!(
                "{}{}{}{}",
                take_str,
                pos_to_string(pos),
                dir_to_string(dir),
                drops_str
            )
        }
    }
}

pub fn ptn_to_action(ptn: &str) -> Option<TakAction> {
    let chars = ptn.trim().chars().collect::<Vec<_>>();
    if chars.is_empty() {
        return None;
    }
    let move_idx = chars
        .iter()
        .position(|&c| c == '+' || c == '-' || c == '<' || c == '>');
    if let Some(move_idx) = move_idx {
        if move_idx < 2 {
            return None;
        }
        let (take_str, rest) = chars.split_at(move_idx - 2);
        let dir = match rest[2] {
            '+' => TakDir::Up,
            '-' => TakDir::Down,
            '<' => TakDir::Left,
            '>' => TakDir::Right,
            _ => return None,
        };
        if rest.len() < 3 {
            return None;
        }
        let pos = parse_pos(&rest[..2])?;
        let take_num = if take_str.is_empty() {
            1
        } else {
            take_str.iter().collect::<String>().parse::<u32>().ok()?
        };
        let drops = rest[3..]
            .iter()
            .map(|c| c.to_digit(10))
            .collect::<Option<Vec<_>>>()?;
        let drops = if drops.len() == 0 {
            vec![take_num]
        } else {
            drops
        };

        Some(TakAction::Move { pos, dir, drops })
    } else {
        if chars.len() == 2 {
            let pos = parse_pos(&chars)?;
            Some(TakAction::Place {
                pos,
                variant: TakVariant::Flat,
            })
        } else if chars.len() == 3 {
            let variant = match chars[0] {
                'S' => TakVariant::Standing,
                'C' => TakVariant::Capstone,
                _ => return None,
            };
            let pos = parse_pos(&chars[1..])?;
            Some(TakAction::Place { pos, variant })
        } else {
            None
        }
    }
}

pub fn game_state_to_string(game_state: &TakGameState) -> String {
    match game_state {
        TakGameState::Ongoing => "0-0".to_string(),
        TakGameState::Win { winner, reason } => {
            let letter = match reason {
                TakWinReason::Road => "R",
                TakWinReason::Flats => "F",
                TakWinReason::Default => "1",
            };
            match winner {
                TakPlayer::White => format!("{}-0", letter),
                TakPlayer::Black => format!("0-{}", letter),
            }
        }
        TakGameState::Draw => "1/2-1/2".to_string(),
    }
}

fn variant_to_string(variant: &TakVariant) -> &'static str {
    match variant {
        TakVariant::Flat => "",
        TakVariant::Standing => "S",
        TakVariant::Capstone => "C",
    }
}

fn pos_to_string(pos: &TakPos) -> String {
    format!("{}{}", (b'a' + pos.x as u8) as char, pos.y + 1)
}

fn dir_to_string(dir: &TakDir) -> &'static str {
    match dir {
        TakDir::Up => "+",
        TakDir::Down => "-",
        TakDir::Left => "<",
        TakDir::Right => ">",
    }
}

fn parse_pos(chars: &[char]) -> Option<TakPos> {
    if chars.len() != 2 {
        return None;
    }
    let x = chars[0];
    let y = chars[1];
    if !('a'..='z').contains(&x) || !('1'..='9').contains(&y) {
        return None;
    }
    Some(TakPos {
        x: (x as u8 - b'a') as i32,
        y: (y as u8 - b'1') as i32,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_action_to_ptn() {
        let action = TakAction::Place {
            pos: TakPos::new(0, 0),
            variant: TakVariant::Flat,
        };
        assert_eq!(action_to_ptn(&action), "a1");
        let action = TakAction::Place {
            pos: TakPos::new(1, 2),
            variant: TakVariant::Standing,
        };
        assert_eq!(action_to_ptn(&action), "Sb3");
        let action = TakAction::Place {
            pos: TakPos::new(2, 4),
            variant: TakVariant::Capstone,
        };
        assert_eq!(action_to_ptn(&action), "Cc5");

        let action = TakAction::Move {
            pos: TakPos::new(3, 3),
            dir: TakDir::Up,
            drops: vec![1],
        };
        assert_eq!(action_to_ptn(&action), "d4+");
        let action = TakAction::Move {
            pos: TakPos::new(4, 4),
            dir: TakDir::Down,
            drops: vec![2],
        };
        assert_eq!(action_to_ptn(&action), "2e5-");
        let action = TakAction::Move {
            pos: TakPos::new(5, 5),
            dir: TakDir::Left,
            drops: vec![1, 2],
        };
        assert_eq!(action_to_ptn(&action), "3f6<12");
        let action = TakAction::Move {
            pos: TakPos::new(6, 6),
            dir: TakDir::Right,
            drops: vec![1, 5, 1],
        };
        assert_eq!(action_to_ptn(&action), "7g7>151");
    }

    #[test]
    fn test_ptn_to_action() {
        helper_test_ptn_to_action(
            "a1",
            Some(TakAction::Place {
                pos: TakPos::new(0, 0),
                variant: TakVariant::Flat,
            }),
        );

        helper_test_ptn_to_action(
            "Sb3",
            Some(TakAction::Place {
                pos: TakPos::new(1, 2),
                variant: TakVariant::Standing,
            }),
        );

        helper_test_ptn_to_action(
            "Cc5",
            Some(TakAction::Place {
                pos: TakPos::new(2, 4),
                variant: TakVariant::Capstone,
            }),
        );

        helper_test_ptn_to_action(
            "d4+",
            Some(TakAction::Move {
                pos: TakPos::new(3, 3),
                dir: TakDir::Up,
                drops: vec![1],
            }),
        );

        helper_test_ptn_to_action(
            "2e5-",
            Some(TakAction::Move {
                pos: TakPos::new(4, 4),
                dir: TakDir::Down,
                drops: vec![2],
            }),
        );

        helper_test_ptn_to_action(
            "3f6<12",
            Some(TakAction::Move {
                pos: TakPos::new(5, 5),
                dir: TakDir::Left,
                drops: vec![1, 2],
            }),
        );

        helper_test_ptn_to_action(
            "7g7>151",
            Some(TakAction::Move {
                pos: TakPos::new(6, 6),
                dir: TakDir::Right,
                drops: vec![1, 5, 1],
            }),
        );

        helper_test_ptn_to_action("Rc2", None);
        helper_test_ptn_to_action("a", None);
        helper_test_ptn_to_action("a0", None);
        helper_test_ptn_to_action("1c", None);
        helper_test_ptn_to_action("2c++", None);
        helper_test_ptn_to_action("Cg7+", None);
    }

    fn helper_test_ptn_to_action(ptn: &str, expected: Option<TakAction>) {
        let action = ptn_to_action(ptn);
        assert_eq!(action, expected, "ptn: {}", ptn);
    }

    #[test]
    fn test_game_state_to_string() {
        let state = TakGameState::Ongoing;
        assert_eq!(game_state_to_string(&state), "0-0");

        let state = TakGameState::Win {
            winner: TakPlayer::White,
            reason: TakWinReason::Road,
        };
        assert_eq!(game_state_to_string(&state), "R-0");

        let state = TakGameState::Win {
            winner: TakPlayer::Black,
            reason: TakWinReason::Flats,
        };
        assert_eq!(game_state_to_string(&state), "0-F");

        let state = TakGameState::Win {
            winner: TakPlayer::White,
            reason: TakWinReason::Default,
        };
        assert_eq!(game_state_to_string(&state), "1-0");

        let state = TakGameState::Draw;
        assert_eq!(game_state_to_string(&state), "1/2-1/2");
    }
}

File: packages/tak-frontend/src/tak-core/ptn.ts
import {
  getDefaultReserve,
  type TakAction,
  type TakActionRecord,
  type TakBaseGameSettings,
  type TakDir,
  type TakGameResult,
  type TakOpening,
  type TakPlayer,
  type TakVariant,
} from '.';
import { TakBaseGame } from './base';
import { TakBoard, type TakStack } from './board';

export function actionFromString(str: string): TakAction | null {
  function stringToVariant(variant: string): TakVariant | null {
    switch (variant) {
      case 'S':
        return 'standing';
      case 'C':
        return 'capstone';
      case 'F':
        return 'flat';
      default:
        return null;
    }
  }
  function stringToDir(dir: string): TakDir | null {
    switch (dir) {
      case '>':
        return 'right';
      case '<':
        return 'left';
      case '+':
        return 'up';
      case '-':
        return 'down';
      default:
        return null;
    }
  }

  const moveRegex = /^([1-9]?)([a-z])([1-9])([<>+-])([1-9]*)/;
  const moveMatch = moveRegex.exec(str);
  if (moveMatch) {
    const take =
      moveMatch[1] !== undefined && moveMatch[1].length > 0
        ? moveMatch[1].charCodeAt(0) - '0'.charCodeAt(0)
        : 1;
    const x = (moveMatch[2] ?? '').charCodeAt(0) - 'a'.charCodeAt(0);
    const y = (moveMatch[3] ?? '').charCodeAt(0) - '1'.charCodeAt(0);
    const dir = stringToDir(moveMatch[4] ?? '');
    if (!dir) {
      return null;
    }
    const drops =
      moveMatch[5] !== undefined && moveMatch[5].length > 0
        ? moveMatch[5].split('').map((d) => d.charCodeAt(0) - '0'.charCodeAt(0))
        : [take];
    return { type: 'move', pos: { x, y }, dir, drops };
  }

  const placeRegex = /^([FSC]?)([a-z])([1-9])/;
  const placeMatch = placeRegex.exec(str);
  if (placeMatch) {
    const variant = stringToVariant((placeMatch[1] ?? 'F') || 'F');
    if (!variant) {
      return null;
    }
    const x = (placeMatch[2] ?? '').charCodeAt(0) - 'a'.charCodeAt(0);
    const y = (placeMatch[3] ?? '').charCodeAt(0) - '1'.charCodeAt(0);
    return { type: 'place', variant, pos: { x, y } };
  }

  return null;
}

export function actionToString(move: TakAction): string {
  if (move.type === 'place') {
    const variant = move.variant === 'capstone' ? 'C' : move.variant === 'standing' ? 'S' : '';

    const col = String.fromCharCode(move.pos.x + 'a'.charCodeAt(0));
    const row = move.pos.y + 1;
    return `${variant}${col}${row.toString()}`;
  } else {
    const col = String.fromCharCode(move.pos.x + 'a'.charCodeAt(0));
    const row = move.pos.y + 1;
    const dir =
      move.dir === 'up' ? '+' : move.dir === 'down' ? '-' : move.dir === 'left' ? '<' : '>';
    const takeSum = move.drops.reduce((a, b) => a + b, 0);
    const take = takeSum === 1 ? '' : takeSum.toString();
    const drops = move.drops.length === 1 ? '' : move.drops.join('');
    return `${take}${col}${row.toString()}${dir}${drops}`;
  }
}

function openingToString(opening: TakOpening): string {
  switch (opening) {
    case 'swap':
      return 'swap';
    case 'noSwap':
      return 'no-swap';
    case 'doubleStack':
      return 'double black stack';
  }
}

function openingFromString(str: string): TakOpening | null {
  switch (str) {
    case 'swap':
      return 'swap';
    case 'no-swap':
      return 'noSwap';
    case 'double black stack':
      return 'doubleStack';
    default:
      return null;
  }
}

export function gameToPTN(
  settings: TakBaseGameSettings,
  history: TakActionRecord[],
  gameState: TakGameResult | null,
  usernames?: Record<TakPlayer, string>,
) {
  const attributes = [
    { name: 'Size', value: settings.boardSize.toString() },
    { name: 'Komi', value: (settings.halfKomi / 2).toString() },
    { name: 'Flats', value: settings.reserve.pieces.toString() },
    { name: 'Caps', value: settings.reserve.capstones.toString() },
    { name: 'Opening', value: openingToString(settings.opening) },
  ];

  if (gameState) {
    attributes.push({ name: 'Result', value: gameResultToString(gameState) });
  }

  if (usernames) {
    attributes.push(
      { name: 'Player1', value: usernames.white },
      { name: 'Player2', value: usernames.black },
    );
  }

  const moves = history.map((record) => actionToString(record.action));

  const movePairs = [];
  for (let i = 0; i < moves.length; i += 2) {
    const firstMove = moves[i] ?? '';
    const secondMove = moves[i + 1] ?? '';
    movePairs.push(firstMove + (secondMove.length > 0 ? ` ${secondMove}` : ''));
  }
  const moveStr = movePairs.map((pair, index) => `${(index + 1).toString()}. ${pair}`).join('\n');

  return `${attributes.map((attr) => `[${attr.name} "${attr.value}"]`).join('\n')}\n${moveStr}`;
}

const PTN_ATTRIBUTES_REGEX = /(?:\[(\w*)\s"([^"]*)"\])\s*/g;
const PTN_MOVES_REGEX = /(?!\d*\.)[^\s]+/g;

const PTN_GAME_OVER_REGEX = /^(1\/2-1\/2|0-1|1-0|0-F|F-0|0-R|R-0|0-0)/;
const INTEGER_REGEX = /^\d+$/;

export function PTNToGame(ptn: string): {
  game: TakBaseGame;
  playerInfo: Record<TakPlayer, { username: string; rating?: number }>;
} | null {
  const attributeMatches = Array.from(ptn.matchAll(PTN_ATTRIBUTES_REGEX)).map((match) => {
    const [, name, value] = match;
    return [name ?? '', value ?? ''] as const;
  });
  const attributes: Map<string, string> = new Map<string, string>(attributeMatches);
  const size = attributes.get('Size');
  const flats = attributes.get('Flats');
  const caps = attributes.get('Caps');
  const komi = attributes.get('Komi');
  const player1 = attributes.get('Player1') ?? 'Player 1';
  const player2 = attributes.get('Player2') ?? 'Player 2';
  const rating1Str = attributes.get('Rating1');
  const rating2Str = attributes.get('Rating2');
  const openingStr = attributes.get('Opening');
  const rating1 = rating1Str !== undefined ? parseInt(rating1Str) : undefined;
  const rating2 = rating2Str !== undefined ? parseInt(rating2Str) : undefined;
  const playerInfo = {
    white: {
      username: player1,
      rating: rating1 !== undefined && !isNaN(rating1) ? rating1 : undefined,
    },
    black: {
      username: player2,
      rating: rating2 !== undefined && !isNaN(rating2) ? rating2 : undefined,
    },
  };
  if (size === undefined || !INTEGER_REGEX.test(size)) return null;
  const sizeNum = parseInt(size);
  const reserve = getDefaultReserve(sizeNum);
  if (flats !== undefined && INTEGER_REGEX.test(flats)) reserve.pieces = parseInt(flats);
  if (caps !== undefined && INTEGER_REGEX.test(caps)) reserve.capstones = parseInt(caps);
  let halfKomi = 0;
  if (komi !== undefined && !Number.isNaN(parseFloat(komi)))
    halfKomi = Math.floor(parseFloat(komi) * 2);

  const restString = ptn.replaceAll(PTN_ATTRIBUTES_REGEX, '').trim();
  const moveMatches = Array.from(restString.matchAll(PTN_MOVES_REGEX));
  const moves = moveMatches.map((match) => match[0]);
  const gameOverStr = moves.find((move) => PTN_GAME_OVER_REGEX.test(move));
  const filteredMoves = moves
    .filter((move) => !PTN_GAME_OVER_REGEX.test(move))
    .map((move) => actionFromString(move.trim()));
  const gameState = gameResultFromString(gameOverStr ?? '');
  const opening = openingFromString(openingStr ?? 'swap');
  if (!opening) {
    console.error('Invalid opening string in PTN:', openingStr);
    return null;
  }

  const gameSettings: TakBaseGameSettings = {
    boardSize: parseInt(size),
    halfKomi,
    reserve,
    opening,
  };
  const game = new TakBaseGame(gameSettings);
  for (const move of filteredMoves) {
    if (!move) {
      console.error('Invalid move string in PTN:', move);
      return null;
    }
    try {
      game.doAction(move);
    } catch (error) {
      console.error('Error applying move:', move, error);
      return null;
    }
  }
  if (gameState && !game.gameResult) {
    game.gameResult = gameState;
  }
  return { game, playerInfo };
}

export function gameResultToString(gameResult: TakGameResult): string {
  switch (gameResult.type) {
    case 'win': {
      const letter = gameResult.reason === 'flats' ? 'F' : gameResult.reason === 'road' ? 'R' : '1';
      return gameResult.winner === 'white' ? `${letter}-0` : `0-${letter}`;
    }
    case 'draw':
      return '1/2-1/2';
    case 'aborted':
      return '0-0';
  }
}

export type TakPTNGameResult = '1-0' | '0-1' | '1/2-1/2' | '0-F' | 'F-0' | '0-R' | 'R-0' | '0-0';

export function gameResultFromString<T extends string>(
  gameOverStr: T,
): T extends TakPTNGameResult ? TakGameResult : TakGameResult | null {
  switch (gameOverStr) {
    case '1/2-1/2':
      return { type: 'draw' };
    case '0-1':
      return { type: 'win', winner: 'black', reason: 'default' };
    case '1-0':
      return { type: 'win', winner: 'white', reason: 'default' };
    case '0-F':
      return { type: 'win', winner: 'black', reason: 'flats' };
    case 'F-0':
      return { type: 'win', winner: 'white', reason: 'flats' };
    case '0-R':
      return { type: 'win', winner: 'black', reason: 'road' };
    case 'R-0':
      return { type: 'win', winner: 'white', reason: 'road' };
    case '0-0':
      return { type: 'aborted' };
    default:
      return null as T extends TakPTNGameResult ? TakGameResult : null;
  }
}

export function gameToTPS(game: TakBaseGame): string {
  const boardStr = boardToPositionString(game.board);
  const playerStr = game.actionHistory.length % 2 === 0 ? '1' : '2';
  const moveStr = (Math.floor(game.actionHistory.length / 2) + 1).toString();
  return `${boardStr} ${playerStr} ${moveStr}`;
}

export function boardToPositionString(board: TakBoard): string {
  function variantToString(variant: TakVariant) {
    switch (variant) {
      case 'flat':
        return '';
      case 'standing':
        return 'S';
      case 'capstone':
        return 'C';
    }
  }

  function rowToPositionString(row: (TakStack | undefined)[]) {
    const result: string[] = [];
    let emptyCount = 0;

    for (const stack of row) {
      if (stack === undefined) {
        emptyCount++;
      } else {
        if (emptyCount > 0) {
          result.push(`x${emptyCount === 1 ? '' : emptyCount.toString()}`);
          emptyCount = 0;
        }
        result.push(
          `${stack.composition
            .map((piece) => (piece.player === 'white' ? '1' : '2'))
            .join('')}${variantToString(stack.variant)}`,
        );
      }
    }

    if (emptyCount > 0) {
      result.push(`x${emptyCount === 1 ? '' : emptyCount.toString()}`);
    }

    return result.join(',');
  }
  const rows = [];
  for (let y = board.size - 1; y >= 0; y--) {
    const row = board.stacks.slice(y * board.size, (y + 1) * board.size);
    rows.push(rowToPositionString(row));
  }
  return rows.join('/');
}
